windows/tools/OobSimulator/MainWindow.xaml.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation and Contributors.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "MainWindow.xaml.h"
+#if __has_include("MainWindow.g.cpp")
+#include "MainWindow.g.cpp"
+#endif
+
+#include <fstream>
+#include "uwb/protocols/fira/UwbConfigurationBuilder.hxx"
+#include "Utils.hxx"
+
+using namespace winrt;
+using namespace Microsoft::UI::Xaml;
+using namespace uwb::protocol::fira;
+
+// To learn more about WinUI, the WinUI project structure,
+// and more about our project templates, see: http://aka.ms/winui-project-info.
+
+namespace winrt::OobSimulator::implementation
+{
+    MainWindow::MainWindow()
+    {
+        InitializeComponent();
+    }
+
+    void MainWindow::SetUwbSessionData(UwbSessionData const& uwbSessionData, std::string const& outputDirectory)
+    {
+        // Convert the input data into JSON
+        nlohmann::json json{};
+        json["sessionId"] = uwbSessionData.sessionId;
+        json["deviceRole"] = static_cast<int>(uwbSessionData.uwbConfiguration.GetDeviceRole().value());
+        json["multiNodeMode"] = static_cast<int>(uwbSessionData.uwbConfiguration.GetMultiNodeMode().value());
+        json["controllerMacAddress"] = uwbSessionData.uwbConfiguration.GetControllerMacAddress().value().ToString();         // Is ToString() the best way to store this?

```

Comment:
Great callout, this is important.

Using `ToString()` has some risk because it's not assumed to be stable over time. The library that exposes `UwbSessionData` should expose either serialize/deserialize functions (eg. , say `ToJson()`, `FromJson()`) or some kind of non-member helper code which serializes/deserializes the type. We've used the latter already with `nlohmann` which supports defining `to_json` and `from_json` functions in the hosting namespace to define serialization/deserialization code. See [`UwbMacAddressJsonSerializer.cxx`](https://github.com/microsoft/nearobject-framework/blob/develop/lib/nearobject/serialization/UwbMacAddressJsonSerializer.cxx) for an example. This way the library itself provides a strong contract that guarantees stability over time which consumers can use and rely on. The reason we preferred the non-member approach is that it separates the concerns of the functionality of the class and the serialization of the class. There's really no reason the class functionality need be concerned about the serialization requirements outside of it.

Coincidentally, most of the types you need here are either primitive or already have `nlohmann` serializers defined. So, I suggest adding an `nlohmann` json serializer for `UwbSessionData`.

Reply:
We did a lot of the service work during the original design. It was mostly just quick and dirty stuff for prototyping and shaping, but a lot of it turned out to be useful as it was.